src/orange/univ/SimpleUnitTracker.hh

```diff
@@ -202,26 +201,27 @@ CELER_FUNCTION auto
 SimpleUnitTracker::cross_boundary(LocalState const& state) const -> Initialization
 {
     CELER_EXPECT(state.surface && state.volume);
-    detail::SenseCalculator calc_senses(
-        this->make_surface_visitor(), state.pos, state.temp_sense);
 
     detail::OnLocalSurface on_surface;
-    auto is_inside = [this, &state, &calc_senses, &on_surface](
-                         LocalVolumeId const& id) -> bool {
+    auto is_inside
+        = [this, &state, &on_surface](LocalVolumeId const& id) -> bool {
         if (id == state.volume)
         {
             // Cannot cross surface into the same volume
             return false;
         }
 
         VolumeView vol = this->make_local_volume(id);
-        auto logic_state
-            = calc_senses(vol, detail::find_face(vol, state.surface));
-
-        if (detail::LogicEvaluator(vol.logic())(logic_state.senses))
+        auto on_face = detail::find_face(vol, state.surface);
+        detail::LazySenseCalculator calc_senses(this->make_surface_visitor(),
```

Comment:
Looking at this more, I don't think we necessarily need the sense calculator to be "lazy and cache", I think we just need to be "on demand" rather than "precalculate". The caching will only help if we've got the same surface showing (and being evaluated!) multiple times within an expression, whereas the benefit of infix notation is using short circuiting so that we don't have to evaluate the contents at all...

Reply:
Initially, I didn't have caching at all. This became necessary when we need to keep track of which senses have been flipped, with the on-demand calculation there's no way to pass that information.

Reply:
Though that cache resides in gmem, so it is very slow... 😅 If we don't need to keep track of flipped sense, then you're right that it is very likely that not doing any caching and not having to clear that global memory every time is faster, assuming recalculating the same sense is rare.

Reply:
Is the "flip" only needed during complex tracking? I thought about this before: https://github.com/celeritas-project/celeritas/issues/1282#issuecomment-2176825606
and if we do re-evaluate senses at every crossing (rather than flipping their internal state and caching them) *except* for the one being crossed (which is analytic) then I think we're ok?

Reply:
yes, this is only done in `complex_intersect`. I'm not very familiar with the logic of why flipping senses is needed, but if it can be avoided it is probably best with the infix evaluation and short-circuiting. We can have a `LazySenseCalculator` without caching and implement a `CachingLazySenseCalculator` as a a front-end to the `LazySenseCalculator`, and provides cached access. Timings would be interesting.

Reply:
@sethrj Yes, it's not ideal but what is currently done in OpenMC is that it moves the particle to an internal surface and then repeats the process of sampling distance to collision and evaluating nearest boundary.

Reply:
Implemented in #1569 and #1576